addons/payment_alipay/models/payment_transaction.py

```diff
@@ -0,0 +1,147 @@
+
+        tx = self.env['payment.transaction'].search([('reference', '=', reference)])
+        if not tx or len(tx) > 1:
+            raise ValidationError(
+                "Alipay: " + _(
+                    "received data with reference %(ref)s matching %(num_tx)d transaction(s)",
+                    ref=reference, num_tx=len(tx)
+                )
+            )
+        return tx
+
+    def _process_feedback_data(self, data):
+        if self.provider != 'alipay':
+            return super()._process_feedback_data(data)
+        if self.state in ['done', 'cancel']:
+            _logger.info(f"Alipay: trying to validate an already terminated tx (ref {self.reference}), "
+                         f"state: {self.state}, received data: {data}")
+            return
+        if float_compare(float(data.get('total_fee', '0.0')), (self.amount + self.fees), 2) != 0:
+            # mc_gross is amount + fees
+            _logger.error(f"Alipay: The paid amount ({data.get('total_fee', '0.0')}) does not match the "
+                          f"total + fees ({self.amount} + {self.fees}) for ref {self.reference}")
+            raise ValidationError("Alipay: " + _("The paid amount does not match the total + fees."))
+        if self.acquirer_id.alipay_payment_method == 'standard_checkout':
+            if data.get('currency') != self.currency_id.name:
+                _logger.error("Alipay: The currency returned by Alipay %(rc)s does not match the "
+                              "transaction currency %(tc)s.)",
+                              rc=data.get('currency'), tc=self.currency_id.name)
+                raise ValidationError("Alipay: " + _("The currency returned by Alipay %(rc)s does not match the "
+                                                     "transaction currency %(tc)s.",
+                                                     rc=data.get('currency'), tc=self.currency_id.name))
+        elif data.get('seller_email') != self.acquirer_id.alipay_seller_email:
+            _logger.error(f"Alipay: The seller email ({data.get('seller_email')}) does not match the configured "
+                          f"Alipay account ({self.acquirer_id.alipay_seller_email}).")
+            raise ValidationError("Alipay: " + _("The seller email does not match the configured Alipay account."))
+
+        status = data.get('trade_status')
+        if status in ['TRADE_FINISHED', 'TRADE_SUCCESS']:
+            self._set_done()
+        elif status == 'TRADE_CLOSED':
+            _logger.info('Received transaction data for Alipay payment %s: set as Canceled' % self.reference)
+            self._set_canceled()
+        else:
+            _logger.info(f'Received unrecognized status for Alipay payment {self.reference}: {status}, set as error')
+            self._set_error("Alipay: " + _(
+                "received data with invalid transaction status: %(tx_status)s", tx_status=status
+            ))
+
+    def _get_specific_rendering_values(self, processing_values):
```

Comment:
Nitpicking: I'd move this one to the top to remain consistent with `payment` and the business flow.